File: 14-TypeScript/Grey Multi-Tenant/packages/grey-vb/Errors/GreyError.vb
' Grey SDK for Visual Basic - Error Class
' Normalized error structure for SDK operations.

Namespace Grey.Errors

    ''' <summary>
    ''' Represents a normalized error from the Grey SDK.
    ''' </summary>
    Public Class GreyError
        Inherits Exception

        ''' <summary>Gets the error code.</summary>
        Public ReadOnly Property Code As String

        ''' <summary>Gets additional error details.</summary>
        Public ReadOnly Property Details As String

        ''' <summary>
        ''' Creates a new GreyError instance.
        ''' </summary>
        ''' <param name="code">The error code.</param>
        ''' <param name="message">The error message.</param>
        ''' <param name="details">Additional error details.</param>
        Public Sub New(code As String, message As String, Optional details As String = Nothing)
            MyBase.New(message)
            Me.Code = code
            Me.Details = details
        End Sub

        ''' <summary>
        ''' Creates a new GreyError with an inner exception.
        ''' </summary>
        Public Sub New(code As String, message As String, innerException As Exception)
            MyBase.New(message, innerException)
            Me.Code = code
            Me.Details = If(innerException?.Message, Nothing)
        End Sub

        ' Factory Methods

        ''' <summary>Creates an unauthorized error.</summary>
        Public Shared Function Unauthorized(Optional message As String = Nothing) As GreyError
            Return New GreyError(
                ErrorCodes.Unauthorized,
                If(message, ErrorCodes.GetMessage(ErrorCodes.Unauthorized)))
        End Function

        ''' <summary>Creates a forbidden error.</summary>
        Public Shared Function Forbidden(Optional message As String = Nothing) As GreyError
            Return New GreyError(
                ErrorCodes.Forbidden,
                If(message, ErrorCodes.GetMessage(ErrorCodes.Forbidden)))
        End Function

        ''' <summary>Creates a not found error.</summary>
        Public Shared Function NotFound(Optional message As String = Nothing) As GreyError
            Return New GreyError(
                ErrorCodes.NotFound,
                If(message, ErrorCodes.GetMessage(ErrorCodes.NotFound)))
        End Function

        ''' <summary>Creates a validation error.</summary>
        Public Shared Function ValidationError(message As String, Optional details As String = Nothing) As GreyError
            Return New GreyError(ErrorCodes.ValidationError, message, details)
        End Function

        ''' <summary>Creates a network error.</summary>
        Public Shared Function NetworkError(Optional message As String = Nothing, Optional innerEx As Exception = Nothing) As GreyError
            If innerEx IsNot Nothing Then
                Return New GreyError(
                    ErrorCodes.NetworkError,
                    If(message, ErrorCodes.GetMessage(ErrorCodes.NetworkError)),
                    innerEx)
            End If
            Return New GreyError(
                ErrorCodes.NetworkError,
                If(message, ErrorCodes.GetMessage(ErrorCodes.NetworkError)))
        End Function

        ''' <summary>Creates a timeout error.</summary>
        Public Shared Function Timeout(Optional message As String = Nothing) As GreyError
            Return New GreyError(
                ErrorCodes.Timeout,
                If(message, ErrorCodes.GetMessage(ErrorCodes.Timeout)))
        End Function

        ''' <summary>Creates a server error.</summary>
        Public Shared Function ServerError(Optional message As String = Nothing, Optional details As String = Nothing) As GreyError
            Return New GreyError(
                ErrorCodes.ServerError,
                If(message, ErrorCodes.GetMessage(ErrorCodes.ServerError)),
                details)
        End Function

        ''' <summary>Creates an error from an HTTP status code.</summary>
        Public Shared Function FromHttpStatus(statusCode As Integer, Optional message As String = Nothing, Optional details As String = Nothing) As GreyError
            Dim code = ErrorCodes.FromHttpStatus(statusCode)
            Return New GreyError(
                code,
                If(message, ErrorCodes.GetMessage(code)),
                details)
        End Function

        ''' <summary>Checks if this error is retryable.</summary>
        Public Function IsRetryable() As Boolean
            Return ErrorCodes.IsRetryable(Code)
        End Function

        ''' <summary>Returns a string representation of the error.</summary>
        Public Overrides Function ToString() As String
            If String.IsNullOrEmpty(Details) Then
                Return $"GreyError [{Code}]: {Message}"
            Else
                Return $"GreyError [{Code}]: {Message} - {Details}"
            End If
        End Function

    End Class

End Namespace

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-r/R/grey_error.R
#' @title Grey Error
#' @description Normalized error structure for the Grey SDK
#' @name grey_error
NULL

#' Create a Grey error object
#'
#' @param code Error code
#' @param message Error message
#' @param details Additional error details (optional)
#' @return A grey_error object (list)
#' @export
grey_error <- function(code, message, details = NULL) {
  err <- list(
    code = normalize_error_code(code),
    message = as.character(message),
    details = details
  )
  class(err) <- c("grey_error", "list")
  err
}

#' Check if an object is a grey_error
#'
#' @param x Object to check
#' @return TRUE if x is a grey_error
#' @export
is_grey_error <- function(x) {
  inherits(x, "grey_error")
}

#' Create an unauthorized error
#'
#' @param message Custom message (optional)
#' @return grey_error
#' @export
error_unauthorized <- function(message = NULL) {
  grey_error(
    ERROR_CODES$UNAUTHORIZED,
    message %||% "Authentication required"
  )
}

#' Create a forbidden error
#'
#' @param message Custom message (optional)
#' @return grey_error
#' @export
error_forbidden <- function(message = NULL) {
  grey_error(
    ERROR_CODES$FORBIDDEN,
    message %||% "Permission denied"
  )
}

#' Create a not found error
#'
#' @param message Custom message (optional)
#' @return grey_error
#' @export
error_not_found <- function(message = NULL) {
  grey_error(
    ERROR_CODES$NOT_FOUND,
    message %||% "Resource not found"
  )
}

#' Create a validation error
#'
#' @param message Error message
#' @param details Validation details (optional)
#' @return grey_error
#' @export
error_validation <- function(message, details = NULL) {
  grey_error(
    ERROR_CODES$VALIDATION_ERROR,
    message,
    details
  )
}

#' Create a network error
#'
#' @param message Custom message (optional)
#' @return grey_error
#' @export
error_network <- function(message = NULL) {
  grey_error(
    ERROR_CODES$NETWORK_ERROR,
    message %||% "Network error occurred"
  )
}

#' Create a timeout error
#'
#' @param message Custom message (optional)
#' @return grey_error
#' @export
error_timeout <- function(message = NULL) {
  grey_error(
    ERROR_CODES$TIMEOUT,
    message %||% "Request timed out"
  )
}

#' Create a server error
#'
#' @param message Custom message (optional)
#' @return grey_error
#' @export
error_server <- function(message = NULL) {
  grey_error(
    ERROR_CODES$SERVER_ERROR,
    message %||% "Server error occurred"
  )
}

#' Create an unknown error
#'
#' @param message Custom message (optional)
#' @return grey_error
#' @export
error_unknown <- function(message = NULL) {
  grey_error(
    ERROR_CODES$UNKNOWN,
    message %||% "An unknown error occurred"
  )
}

#' Create error from HTTP response
#'
#' @param status HTTP status code
#' @param body Response body (list)
#' @return grey_error
#' @export
error_from_response <- function(status, body = NULL) {
  code <- error_code_from_status(status)
  message <- body$message %||% body$error %||% paste("HTTP error", status)
  details <- body$details
  
  grey_error(code, message, details)
}

#' Null-coalescing operator
#' @param a First value
#' @param b Default value if a is NULL
#' @return a if not NULL, otherwise b
`%||%` <- function(a, b) {
  if (is.null(a)) b else a
}
